### PR TITLE
initial tracing setup

### PR DESCRIPTION
**Pull request by erinborders**
Adding distributed tracing handler

### CODE REVIEW

File: v2/processor.go [hunk 1]
```diff
@@ -118,6 +120,43 @@ func NewPanicHandler(handler HandlerFunc) HandlerFunc {
...
+	return func(ctx context.Context, settler MessageSettler, message *azservicebus.ReceivedMessage) {
+		var err error
+		if message != nil && message.ApplicationProperties != nil {
+			ctx, err = ExtractMessageContext(message)
+			if err != nil {
+				// do something
+			}
+		}
+
+		ctx, span := tab.StartSpan(ctx, "")
```
Comment (serbrech) [inline v2/processor.go:134]: need to define a naming prefix for go-shuttle spans and keep it consistent.

File: v2/processor.go [hunk 2]
```diff
@@ -118,6 +120,43 @@ func NewPanicHandler(handler HandlerFunc) HandlerFunc {
...
+				tab.StringAttribute("message.id", message.MessageID),
+				tab.StringAttribute("message.correlationId", *message.CorrelationID))
+			if message.ScheduledEnqueueTime != nil {
+				span.AddAttributes(tab.StringAttribute("message.scheduledEnqueuedTime", message.ScheduledEnqueueTime.String()))
+			}
+			if message.TimeToLive != nil {
+				span.AddAttributes(tab.StringAttribute("message.ttl", message.TimeToLive.String()))
+			}
+		} else {
+			span.Logger().Error(fmt.Errorf("message is nil"))
```
Comment (serbrech) [inline v2/processor.go:146]: Maybe Warn, instead of Error.

Comment (erinborders) [reply v2/processor.go]: Okay, I didn't see a warn method for the span logger so I used info and just included warning in the message

File: v2/processor.go (hunk 1 above)
Comment (erinborders) [reply v2/processor.go]: Ok made the change - I modeled the naming prefix off what I saw in the original go-shuttle, what do you think?

File: v2/processor.go [hunk 3]
```diff
@@ -169,3 +199,20 @@ func (plr *peekLockRenewer) startPeriodicRenewal(ctx context.Context, message *a
...
 	}
 }
+
+type MessageWrapper struct {
+	Message *azservicebus.ReceivedMessage
+}
+
+// Set implements tab.Carrier interface
+func (mw *MessageWrapper) Set(key string, value interface{}) {
+	if mw.Message.ApplicationProperties == nil {
```
Comment (serbrech) [inline v2/processor.go:209]: check if Message is nil too. we don't want this code to cause panics in user's message handling

File: v2/processor.go [hunk 4]
```diff
@@ -118,6 +119,35 @@ func NewPanicHandler(handler HandlerFunc) HandlerFunc {
 	}
 }
 
+// NewTracingHandler extracts the context from the message Application property if available, or from the existing
```
Comment (serbrech) [inline v2/processor.go:122]: Could you move the tracing handler to v2/tracing/tab?

**Comment by serbrech** [comment]
this looks good :)
I pushed a few small changes that caused minor conflicts though

File: v2/processor.go [hunk 5]
```diff
@@ -184,6 +183,28 @@ func (plr *peekLockRenewer) startPeriodicRenewal(ctx context.Context, message *a
 	}
 }
 
+type MessageWrapper struct {
```
Comment (serbrech) [inline v2/processor.go:186]: these should move to the tracing package and be private.

**Review by serbrech — APPROVED** [review]

**Pull request merged**